src/backend/abuseDetection.ts

```diff
@@ -0,0 +1,140 @@
+import * as AWS from 'aws-sdk';
+import { dropRight, flatMap, mapValues, range } from 'lodash';
+import { HOUR_IN_MS } from '../common/time';
+import { nonNullable } from '../common/types';
+
+const TIME_RANGE_HOURS = 24;
+
+export type AbuseFingerprint = {
+  source_ip: string;
+  user_agent: string;
+  forwarded_for: string;
+};
+export type AbuseScore = {
+  [key in keyof AbuseFingerprint]: number;
+};
+
+const emptyScore: AbuseScore = { source_ip: 0, user_agent: 0, forwarded_for: 0 };
+const fingerprintKeys = (Object.keys(emptyScore) as any) as Array<keyof AbuseFingerprint>;
+
+export const ABUSE_SCORE_ERROR: AbuseScore = { source_ip: -2, user_agent: -2, forwarded_for: -2 }; // this special score can be used to indicate that a given response couldn't be scored, due to e.g. DynamoDB error or some such
+export const ABUSE_SCORE_MISSING = -1; // this special score can be used to indicate that because there was no input value for something like X-Forwarded-For, there's no point in scoring it for this request
+
+// Reads & updates DynamoDB to keep track of our abuse detection metrics.
+// Promises the "abuse score" for the given fingerprint (i.e. request).
+// The read/write Promises are returned separately, so that client requests can be serviced immediately after reading the current abuse score finishes.
+// Thus the client doesn't need to wait for the writes to finish; those can happen after the client's already gotten its response.
+export function performAbuseDetection(
+  dynamoDb: DynamoDBClient,
+  fingerprint: AbuseFingerprint,
+  hashFunction: (input: string) => string,
+  // Allow overriding defaults in test code:
+  timestamp = Date.now,
+  timeRangeHours = TIME_RANGE_HOURS,
+) {
+  const now = timestamp();
+  const timeRange = getTimeRange(now, timeRangeHours);
+  const fp = mapValues(fingerprint, v => (v ? hashFunction(v) : '')); // to preserve privacy, put all values (e.g. IP, UA) through a one-way hash function before doing anything with them (keep empty strings as empty, though, since they warrant special treatment later)
+  const keysToGet = flatMap(fingerprintKeys, key => timeRange.map(ts => getStorageKey(key, fp[key], ts)));
+  const keysToIncrement = fingerprintKeys
+    .map(key => (fp[key] ? getStorageKey(key, fp[key], now) : null))
+    .filter(nonNullable);
```

Comment:
Tricks of the trade: You can use `Boolean` to do this unless you want to explicitly filter only `null` and `undefined` values.

Reply:
Oh, interesting!

Though after trying this, `const keysToIncrement` is inferred as `(string | null)[]` whereas it's supposed to be just `string[]`. :thinking:

Reply:
That's strange, I tried that on Typescript Playground to assert my claim and got `string[]` out of the `.filter` 🧐